### PR TITLE
Add resource option PreCheckCallback

### PR DESCRIPTION
**Pull request by iwahbe**
Add the ability to override the check behavior for resources.

### CODE REVIEW

**Comment by t0yv0** [comment]
https://go.dev/play/p/-eNaTRqOVhu perhaps consider a way to override it without making changes to this codebase? It seems Go language has the facility to do this via embedded structs.

**Review by t0yv0 — CHANGES_REQUESTED** [review]
Can it not be overridden in Go by other means outside of making edits to this codebase?

**Comment by t0yv0** [comment]
IN addition, FWIW there's PreConfigureCallback machinery that is allowed to modify the data at Pulumi layer before it drops to TF; not sure if that is sufficient for the use case you have.

**Comment by iwahbe** [comment]
> https://go.dev/play/p/-eNaTRqOVhu perhaps consider a way to override it without making changes to this codebase? It seems Go language has the facility to do this via embedded structs.

I don't see how this is applicable to this case. Where would this come into play? Which interface would I be hijacking here? 

`PreConfigureCallback` is for provider configuration, not resource configuration. It is not sufficient for my use case.

**Comment by t0yv0** [comment]
Ok, my bad you're right about PreConfigureCallback belonging to a different part of the system.

Regarding overriding via Go embedding I guess it can be too tricky to be useful if we rely on calling into Main/ having multiple layers.

So let's discuss making a change to this codebase to support customizing Check. One sec.

File: pf/tfbridge/provider_check.go [hunk 1]
```diff
@@ -46,6 +46,15 @@ func (p *provider) CheckWithContext(
 		return checkedInputs, []plugin.CheckFailure{}, err
 	}
 
+	if info := rh.pulumiResourceInfo; info != nil {
```
Comment (t0yv0) [inline pf/tfbridge/provider_check.go:49]: This bypasses defaults.ApplyDefaultInfoValues and p.validateResourceConfig and tfbridge.MarkSchemaSecrets potentially dropping the features implemented in this code. Would it be worthwhile to figure out how to express the transformation in a way that still calls into that code and keeps the features? Perhaps something along the lines of PreConfigureCallback like PreCheckCallback?

Comment (t0yv0) [reply pf/tfbridge/provider_check.go:49]: Ah never mind. It keeps those in place.

**Comment by t0yv0** [comment]
Giving it a second read, I think this can be useful to have, I wanted to reach out for something like this in other work. Perhaps let's call it PreCheckCallback? Add some quick tests to make sure it's covered, similar to PreConfigureCallback.

File: pf/tfbridge/provider_check.go [hunk 2]
```diff
@@ -46,6 +46,15 @@ func (p *provider) CheckWithContext(
...
+		if check := info.XCustomCheck; check != nil {
+			checkedInputs, err := check(ctx, checkedInputs, p.lastKnownProviderConfig.Copy())
```
[the remark this answers is not in the document]
Comment (iwahbe) [reply pf/tfbridge/provider_check.go:51]: Yep. I would like this to see values as passed in, not as configured.

[the remark this answers is not in the document]
Comment (t0yv0) [reply pf/tfbridge/provider_check.go:51]: Perfect, makes sense.

**Comment by t0yv0** [comment]
This is great just needs a little more test TLC.

File: pkg/tfbridge/provider_test.go [hunk 3]
```diff
@@ -550,6 +521,104 @@ func TestProviderCheckV2(t *testing.T) {
...
+	callback := func(
+		_ context.Context, config, meta resource.PropertyMap,
+	) (resource.PropertyMap, error) {
+		config["arrayPropertyValues"] = resource.NewArrayProperty(
+			[]resource.PropertyValue{meta["prop"]},
+		)
+		return config, nil
+	}
+
+	for _, f := range factories {
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:601]: By-reference loop variable capture :)

File: pkg/tfbridge/provider_test.go [hunk 4]
```diff
@@ -550,6 +521,104 @@ func TestProviderCheckV2(t *testing.T) {
...
+			    "request": {
+			      "urn": "urn:pulumi:st::pg::testprovider:index/res:Res::r",
+			      "olds": {},
+			      "news": {},
+			      "randomSeed": "wqZZaHWVfsS1ozo3bdauTfZmjslvWcZpUjn7BzpS79c="
+			    },
+			    "response": {
+			      "inputs": {
+				"__defaults": [],
+				"arrayPropertyValues": ["global"]
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:585]: Thank you!

File: pkg/tfbridge/provider_test.go (hunk 3 above)
Comment (iwahbe) [reply pkg/tfbridge/provider_test.go:602]: The test isn't parallelized. I can change that (https://github.com/pulumi/pulumi-terraform-bridge/commit/272e881f24744a86655da45c59f5c903d5c0de46), but the by-ref isn't actually a risk here.

Comment (t0yv0) [reply pkg/tfbridge/provider_test.go:602]: Yeah I wanted to enable a linter that flags these but we can deal with it then.

Comment (iwahbe) [reply pkg/tfbridge/provider_test.go:602]: Have you seen https://github.com/golang/go/discussions/56010?

**Pull request merged**